### PR TITLE
[onert] Add quant8 support for Tanh cker

### PR DESCRIPTION
**Pull request by hasw7569**
Add quantization support for Tanh cker

ONE-DCO-1.0-Signed-off-by: Leehansol <hansolu.lee@samsung.com>

### CODE REVIEW

File: compute/cker/include/cker/operation/Tanh.h [hunk 1]
```diff
@@ -36,6 +37,40 @@ inline void Tanh(const Shape &input_shape, const float *input_data, const Shape
   output_map.array() = input_map.array().tanh();
 }
 
+inline void Quant8_tanh(int32_t input_zero_point, int32_t input_range_radius,
+                        int32_t input_multiplier, int32_t input_shift,
+                        int32_t input_size, const int8_t *input_data,
+                        int8_t *output_data)
+{
+  static constexpr int32_t kInputIntegerBits = 4;
```
Comment (dr-venkman) [inline compute/cker/include/cker/operation/Tanh.h:45]: This implementation appears to come from tflite `reference_interger_ops`, which is for `int16_t`. The implementation from `optimized_ops` needs to be ported. Could you please check once?

File: runtime/onert/backend/cpu/ops/TanhLayer.cc [hunk 2]
```diff
@@ -42,14 +42,37 @@ void TanhLayer::tanhFloat32()
 
 void TanhLayer::tanhQuant8()
 {
-  // cker quant8 tanh is not implemented yet
-  throw std::runtime_error{"NYI"};
+  switch (_output->data_type())
+  {
+    case OperandType::QUANT_UINT8_ASYMM:
```
Comment (dr-venkman) [inline runtime/onert/backend/cpu/ops/TanhLayer.cc:47]: Do we need to type check here, or can it be done at an earlier phase?

File: compute/cker/include/cker/operation/Tanh.h (hunk 1 above)
Comment (hasw7569) [reply compute/cker/include/cker/operation/Tanh.h:45]: I think tensorflow has two implementations of tanh function to support int8 type.

the first one is placed in tensorflow/lite/kernels/activation.cc -> Tanhval function

the second is placed in tensorflow/lite/micro/kernels/tanh.cc -> Tanhval function

I don't know what is diffrence of first and second.

if you know about it, could you explain diffrence between them ?

Comment (hasw7569) [reply compute/cker/include/cker/operation/Tanh.h:45]: I guess the difference between first and second is the existence of a lookup table.

**Comment by hseok-oh** [comment]
@hasw7569 LGTM. Please update skiplist.

File: runtime/onert/backend/cpu/ops/TanhLayer.cc [hunk 3]
```diff
@@ -34,6 +34,25 @@ TanhLayer::TanhLayer() : _input(nullptr), _output(nullptr)
...
+  const auto input_zero_point = static_cast<int32_t>(_input->data_offset());
+  const auto output_scale = static_cast<double>(_output->data_scale());
+  const auto output_zero_point = static_cast<int32_t>(_output->data_offset());
+  const float inverse_scale = 1 / output_scale;
+  int32_t maxval = std::numeric_limits<uint8_t>::max();
+  int32_t minval = std::numeric_limits<uint8_t>::min();
+  for (int32_t val = minval; val <= maxval; ++val)
+  {
+    const float dequantized = input_scale * (val - input_zero_point);
+    const float transformed = 1.0f / (1.0f + std::exp(-dequantized));
```
Comment (dr-venkman) [inline runtime/onert/backend/cpu/ops/TanhLayer.cc:49]: Please check formula, above expression seems like sigmoid than tanh.

Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/TanhLayer.cc]: Oops..
Thanks for your comment !

**Comment by hasw7569** [comment]
@dr-venkman PTAL

**Comment by hseok-oh** [comment]
@dr-venkman PTAL

**Pull request merged**